crypto/rand/randfile.c

```diff
@@ -1,256 +1,89 @@
-    setbuf(in, NULL);
-
-    for (;;) {
-        if (bytes > 0)
-            n = (bytes <= RAND_LOAD_BUF_SIZE) ? (int)bytes : RAND_BUF_SIZE;
-        else
-            n = RAND_LOAD_BUF_SIZE;
-        i = fread(buf, 1, n, in);
-#ifdef EINTR
-        if (ferror(in) && errno == EINTR) {
-            clearerr(in);
-            if (i == 0)
-                continue;
-        }
-#endif
-        if (i == 0)
-            break;
-
-        RAND_add(buf, i, (double)i);
-        ret += i;
-
-        /* If given a bytecount, and we did it, break. */
-        if (bytes > 0 && (bytes -= i) <= 0)
-            break;
-    }
-
-    OPENSSL_cleanse(buf, sizeof(buf));
-    fclose(in);
-    if (!RAND_status()) {
-        ERR_raise_data(ERR_LIB_RAND, RAND_R_RESEED_ERROR, "Filename=%s", file);
-        return -1;
-    }
-
-    return ret;
+    /* -1 means read the whole file, else read the specified part. */
+    return bytes > 0 ? (int)bytes : RAND_BUF_SIZE;
 }
 
 int RAND_write_file(const char *file)
 {
-    unsigned char buf[RAND_BUF_SIZE];
-    int ret = -1;
-    FILE *out = NULL;
-#ifndef OPENSSL_NO_POSIX_IO
+    /* Only write to plain files.  TOCTOU issue. */
     struct stat sb;
-
-    if (stat(file, &sb) >= 0 && !S_ISREG(sb.st_mode)) {
-        ERR_raise_data(ERR_LIB_RAND, RAND_R_NOT_A_REGULAR_FILE,
-                       "Filename=%s", file);
+    if (stat(file, &sb) < 0 || !S_ISREG(sb.st_mode))
         return -1;
-    }
-#endif
 
-    /* Collect enough random data. */
-    if (RAND_priv_bytes(buf, (int)sizeof(buf)) != 1)
-        return  -1;
+    int fd = open(file, O_WRONLY|O_CREAT, 0600);
+    if (fd < 0)
+        return -1;
 
-#if defined(O_CREAT) && !defined(OPENSSL_NO_POSIX_IO) && \
-    !defined(OPENSSL_SYS_WINDOWS)
-    {
-# ifndef O_BINARY
-#  define O_BINARY 0
-# endif
-        /*
-         * chmod(..., 0600) is too late to protect the file, permissions
-         * should be restrictive from the start
-         */
-        int fd = open(file, O_WRONLY | O_CREAT | O_BINARY, 0600);
-        if (fd != -1)
-            out = fdopen(fd, "wb");
+    unsigned char buff[RAND_BUF_SIZE];
+    if (RAND_bytes(buff, sizeof buff) != 1) {
+        close(fd);
+        return -1;
     }
-#endif
 
-    if (out == NULL)
-        out = openssl_fopen(file, "wb");
+    FILE *out = fdopen(fd, "wb");
     if (out == NULL) {
-        ERR_raise_data(ERR_LIB_RAND, RAND_R_CANNOT_OPEN_FILE,
-                       "Filename=%s", file);
+        close(fd);
         return -1;
     }
 
-#if !defined(NO_CHMOD) && !defined(OPENSSL_NO_POSIX_IO)
-    /*
-     * Yes it's late to do this (see above comment), but better than nothing.
-     */
-    chmod(file, 0600);
-#endif
+    if (fwrite(buff, sizeof buff, 1, out) != sizeof buff) {
+        fclose(out);
+        return -1;
+    }
 
-    ret = fwrite(buf, 1, RAND_BUF_SIZE, out);
     fclose(out);
-    OPENSSL_cleanse(buf, RAND_BUF_SIZE);
-    return ret;
+    OPENSSL_cleanse(buff, sizeof buff);
+    return (int)sizeof buff;;
+
```

Comment:
```suggestion
    return (int)sizeof buff;
```
There's an extra empty line and an extra semicolon here.